x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/X-ValuesAggregator.java.st

```diff
@@ -268,63 +269,131 @@ $endif$
             blocks[offset] = toBlock(driverContext.blockFactory(), selected);
         }
 
+        /**
+         * Builds a {@link Block} with the unique values collected for the {@code #selected}
+         * groups. This is the implementation of the final and intermediate results of the agg.
+         */
         Block toBlock(BlockFactory blockFactory, IntVector selected) {
             if (values.size() == 0) {
                 return blockFactory.newConstantNullBlock(selected.getPositionCount());
             }
-$if(BytesRef)$
-            BytesRef scratch = new BytesRef();
+
+            long selectedCountsSize = 0;
+            long idsSize = 0;
+            try {
+                /*
+                 * Get a count of all groups less than the maximum selected group. Count
+                 * *downwards* so that we can flip the sign on all of the actually selected
+                 * groups. Negative values in this array are always unselected groups.
+                 */
+                int selectedCountsLen = selected.max() + 1;
+                selectedCountsSize = RamUsageEstimator.alignObjectSize(
+                    RamUsageEstimator.NUM_BYTES_ARRAY_HEADER + selectedCountsLen * Integer.BYTES
+                );
+                blockFactory.adjustBreaker(selectedCountsSize);
```

Comment:
I think we should only increase selectedCountsSize after we have successfully requested memory; otherwise, we might incorrectly return memory that we haven't actually acquired.

Reply:
:+1:

I wonder why my tests hadn't caught that.

Reply:
This catches it:
```
while g x-pack:plugin:esql:compute:test -Dtests.class='*Values*Agg*Tests' -Dtests.iters=1000 -Dtests.method="*Cranky*" --fail-fast; do echo ok; done
```
Thanks!